src/librbd/crypto/CryptoObjectDispatch.cc

```diff
@@ -0,0 +1,197 @@
+
+    I* image_ctx;
+    uint64_t object_no;
+    uint64_t object_off;
+    ceph::bufferlist* read_data;
+    Context* onfinish;
+    io::ReadResult::C_ObjectReadRequest* req_comp;
+    io::ObjectDispatchSpec* req;
+
+    C_EncryptedObjectReadRequest(
+            I* image_ctx, uint64_t object_no, uint64_t object_off,
+            uint64_t object_len, librados::snap_t snap_id, int op_flags,
+            const ZTracer::Trace &parent_trace, ceph::bufferlist* read_data,
+            int* object_dispatch_flags, Context** on_finish,
+            Context* on_dispatched) : image_ctx(image_ctx),
+                                      object_no(object_no),
+                                      object_off(object_off),
+                                      read_data(read_data),
+                                      onfinish(on_dispatched) {
+      *on_finish = util::create_context_callback<
+              Context, &Context::complete>(*on_finish, image_ctx->crypto);
+
+      auto aio_comp = io::AioCompletion::create_and_start(
+              (Context*)this, util::get_image_ctx(image_ctx),
+              io::AIO_TYPE_READ);
+      aio_comp->read_result = io::ReadResult{read_data};
+      aio_comp->set_request_count(1);
+
+      auto req_comp = new io::ReadResult::C_ObjectReadRequest(
+              aio_comp, object_off, object_len, {{0, object_len}});
+
+      req = io::ObjectDispatchSpec::create_read(
+              image_ctx, io::OBJECT_DISPATCH_LAYER_CRYPTO, object_no,
+              object_off, object_len, snap_id, op_flags, parent_trace,
+              &req_comp->bl, &req_comp->extent_map, req_comp);
+    }
+
+    void send() {
+      req->send();
+    }
+
+    void finish(int r) override {
+      ldout(image_ctx->cct, 20) << "r=" << r << dendl;
+      if (r > 0) {
+        image_ctx->crypto->decrypt(
+                std::move(*read_data),
+                CryptoObjectDispatch<I>::get_image_offset(
+                        image_ctx, object_no, object_off));
+      }
+      onfinish->complete(r);
+    }
+};
+
+template <typename I>
+CryptoObjectDispatch<I>::CryptoObjectDispatch(
+    I* image_ctx)
+  : m_image_ctx(image_ctx) {
+  ceph_assert(m_image_ctx->crypto != nullptr);
+}
+
+template <typename I>
+void CryptoObjectDispatch<I>::shut_down(Context* on_finish) {
+  on_finish->complete(0);
+}
+
+template <typename I>
+bool CryptoObjectDispatch<I>::read(
+    uint64_t object_no, uint64_t object_off, uint64_t object_len,
+    librados::snap_t snap_id, int op_flags, const ZTracer::Trace &parent_trace,
+    ceph::bufferlist* read_data, io::ExtentMap* extent_map,
+    int* object_dispatch_flags, io::DispatchResult* dispatch_result,
+    Context** on_finish, Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << data_object_name(m_image_ctx, object_no) << " "
+                 << object_off << "~" << object_len << dendl;
+
+  *dispatch_result = io::DISPATCH_RESULT_COMPLETE;
+  auto req = new C_EncryptedObjectReadRequest(
```

Comment:
Why not just let the object request proceed to the next layer and intercept the result for decryption? You can wrap the provided `on_finish` handler so that you can get a callback when it finishes the read. It won't work for the parent-read case, but neither will this approach currently.

Reply:
Well I have already written it that way to prepare for alignment, where you need to modify the original read request.
Since I already have this code, I prefer to avoid writing throw-away code.

Reply:
All the ops will need to handle alignment crypto block alignment, correct? write, write-same, and compare-and-write might need to perform a read-modify-write cycle. And discard would actually need to write encrypted zeroes(?) or perhaps just turn it into a no-op if you are going to store integrity.

Reply:
Yep.
I actually mistaken Discard to be a cache operation, but I now see that it actually is writing zeros, right? (hopefully claiming back OSD space, right?)

Reply:
It's essentially a hint back to the OSD to reclaim space if possible. We (librbd) try to avoid issuing an actual "zero"-ing write operation on a discard and instead just treat the discard as a hint and ignore any extents that will cause us not to reclaim space.